File: src/main/java/io/humble/video/io/URLProtocolManager.java
/*******************************************************************************
 * Copyright (c) 2013, Art Clarke.  All rights reserved.
 *  
 * This file is part of Humble-Video.
 *
 * Humble-Video is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Humble-Video is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Humble-Video.  If not, see <http://www.gnu.org/licenses/>.
 *******************************************************************************/

package io.humble.video.io;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


import io.humble.video.io.FfmpegIO;
import io.humble.video.io.FileProtocolHandlerFactory;
import io.humble.video.io.IURLProtocolHandler;
import io.humble.video.io.IURLProtocolHandlerFactory;
import io.humble.video.io.NullProtocolHandlerFactory;
import io.humble.video.io.URLProtocolManager;

/**
 * This class manages URL protocols that we have handlers for.  It is used
 * (and called from) the Native code that integrates with FFMPEG.
 * 
 * To register a handler you would use the singleton manager as follows:
 * <code>
 * <pre>
 *  URLProtocolManager mgr = URLProtocolManager.getManager();
 *  mgr.registerFactory("myprotocol", myProtocolHandlerFactory);
 * </pre>
 * </code>
 */
public class URLProtocolManager
{
  private final Logger log = LoggerFactory.getLogger(this.getClass());
  private final ConcurrentMap<String, IURLProtocolHandlerFactory> mProtocols;

  public static final String DEFAULT_PROTOCOL = "humblefile";
  public static final String NULL_PROTOCOL = "humblenull";
  
  private static final URLProtocolManager sManager = new URLProtocolManager();

  /**
   * Factory method to get the global protocol manager.
   */
  public static URLProtocolManager getManager()
  {
    return sManager;
  }

  /**
   * Method to be called to initialize this library.
   */
  public static void init()
  {
    // just calling this method causes sManager to be initialized
    // in the static constructor, which is the only initialization
    // we currently need.
  }

  // Don't let people create one of these; instead they must
  // use the Singleton factory method.
  private URLProtocolManager()
  {
    mProtocols = new ConcurrentHashMap<String, IURLProtocolHandlerFactory>();
    // Always register the DEFAULT
    registerFactory(DEFAULT_PROTOCOL, new FileProtocolHandlerFactory());
    // And the NULL protocols
    registerFactory(NULL_PROTOCOL, new NullProtocolHandlerFactory());        
  }

  /**
   * Register a new factory for IURLProtocolHandlers for a given protocol.
   * <p>
   * FFMPEG is very picky; protocols must be only alpha characters (no numbers).
   * </p>
   * @param protocol The protocol, without colon, this factory is to be used for.
   * @param factory The factory for the manager to use whenever a handler is requested for a registered protocol;
   *   null means disable handling for this factory.
   * @return The factory that was previously used for that protocol (or null if none).
   */
  public IURLProtocolHandlerFactory registerFactory(String protocol,
      IURLProtocolHandlerFactory factory)
  {
    if (protocol == null)
      throw new IllegalArgumentException("protocol required");
    
    IURLProtocolHandlerFactory oldFactory;
    if (factory == null)
      oldFactory = mProtocols.remove(protocol);
    else
      oldFactory = mProtocols.put(protocol, factory);
    
    log.trace("Registering factory for URLProtocol: {}", protocol);

    if (oldFactory == null)
    {
      // we previously didn't have something registered for this factory
      // so tell FFMPEG we're now the protocol manager for this protocol.
      log.trace("Letting FFMPEG know about an additional protocol: {}",
          protocol);
      FfmpegIO.registerProtocolHandler(protocol, this);
    }
    return oldFactory;
  }

  /*
   * Get a IURLProtocolHandler for this url.
   * <p>
   * IMPORTANT: This function is called from native code, and so the name and signature cannot
   * change without changing the native code.
   * </p><p>
   * This function is eventually invoked whenever someone tries to call url_open("yourprotocol:...", flags)
   * from FFMPEG native code.  It returns a protocol handler which will then have open(...) called on it.
   * </p>
   * @param url The URL we want to handle.
   * @param flags   Any flags that the url_open() function will want to pass.
   * 
   * @returns A protocol handler to use.
   */
  public IURLProtocolHandler getHandler(String url, int flags)
  {
    IURLProtocolHandler result = null;

    log.trace("looking for protocol handler for: {}", url);
    if (url == null || url.length() == 0)
      throw new IllegalArgumentException("expected valid URL");
    int colonIndex = url.indexOf(":");
    String protocol = null;
    if (colonIndex > 0)
    {
      protocol = url.substring(0, colonIndex);
    }
    else
    {
      protocol = DEFAULT_PROTOCOL;
    }

    IURLProtocolHandlerFactory factory = mProtocols.get(protocol);
    if (factory != null)
    {
      result = factory.getHandler(protocol, url, flags);
    }
    else
    {
      log.error("asked to get handler for unsupported protocol: {}",
          protocol);
    }

    return result;
  }
  
  /**
   * Get the resource portion of a url.  For example for the URL
   * <pre>
   * http://www.humble.io/video
   * </pre>
   * The protocol string is <code>http</code> and the resource
   * string is <code>www.humble.io/video</code>
   * 
   * @param url The url to parse
   * @return The resource
   */
  
  public static String getResourceFromURL(String url)
  {
    String retval = url;
    if (url != null && url.length() > 0)
    {
      int colonIndex;
      colonIndex = url.indexOf("://");
      if (colonIndex > 0)
        retval = url.substring(colonIndex+3);
      else {
        colonIndex = url.indexOf(":");
        if (colonIndex > 0)
        {
          // remove the URL prefix
          retval = url.substring(colonIndex + 1);
        }
      }
    }
    return retval;
  }

  /**
   * Get the protocol portion of a url.  For example for the URL
   * <pre>
   * http://www.humble.io/video
   * </pre>
   * The protocol string is <code>http</code> and the resource
   * string is <code>//www.humble.io/video</code>
   * 
   * @param url The url to parse
   * @return The protocol
   */
  
  public static String getProtocolFromURL(String url)
  {
    String retval = null;
    if (url != null && url.length() > 0)
    {
      int colonIndex = url.indexOf(":");
      if (colonIndex > 0)
      {
        // remove the URL suffix
        retval = url.substring(0, colonIndex);
      }
    }
    return retval;
  }

}

File: humble-video-noarch/src/main/gnu/src/io/humble/video/customio/URLProtocolManager.cpp
/*
 * Copyright (c) 2013-Forward, Andrew "Art" Clarke
 *
 * This file is part of Humble Video.
 * 
 * Humble Video is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Humble Video is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with Humble Video.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Container.h
 *
 *  Created on: Jun 30, 2013
 *      Author: aclarke
 */

#include <exception>
#include <stdexcept>
#include <cstring>
#include <io/humble/video/customio/URLProtocolManager.h>

namespace io { namespace humble { namespace video { namespace customio
{
using namespace std;


static const char*
URLProtocolManager_VALID_PROTOCOL_CHARS =
    "abcdefghijklmnopqrstuvwxyz"
    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    "0123456789+-.";

const char *
URLProtocolManager::parseProtocol(
    char * dest,
    int32_t dest_len,
    const char* url)
{
  if (!dest || dest_len <=0 || !url || !*url)
    return 0;
  size_t url_len = strlen(url);
  if (url_len <= 1)
    return 0;

  size_t proto_len = strspn(url, URLProtocolManager_VALID_PROTOCOL_CHARS);
  if (proto_len <= 0)
    return 0;

  // file protocol
  if (url[proto_len] != ':' && url[proto_len] != ',') {
    return 0;
  }
  if (url[0] && url[1] == ':') {
    return 0;
  }
  if (proto_len >= (size_t)dest_len)
    // not enough space
    return 0;
  strncpy(dest, url, proto_len);
  dest[proto_len] = 0;

  return dest;
}

URLProtocolManager* URLProtocolManager :: mFirstProtocol = 0;

URLProtocolManager*
URLProtocolManager :: getNextProtocol(URLProtocolManager *prev)
{
  return prev ? prev->mNext : mFirstProtocol;
}

URLProtocolManager*
URLProtocolManager :: registerProtocol(URLProtocolManager* mgr)
{
  if (mgr) {
    URLProtocolManager**p;
    p = &mFirstProtocol;
    while(*p) p = &(*p)->mNext;
    *p = mgr;
  }
  return mgr;
}

int
URLProtocolManager :: unregisterAllProtocols()
{
  int total = 0;
  URLProtocolManager *p=mFirstProtocol;
  while (p) {
    URLProtocolManager* casualty = p;
    p = p->mNext;
    delete casualty;
    ++total;
  }
  mFirstProtocol=0;
  return total;
}

URLProtocolManager*
URLProtocolManager :: unregisterProtocol(URLProtocolManager * mgr)
{
  if (mgr) {
    URLProtocolManager **p;
    p = &mFirstProtocol;
    while(*p && *p != mgr)
      p = &(*p)->mNext;
    if (*p)
      *p = mgr->mNext;
    mgr->mNext = 0;
  }
  return mgr;
}

URLProtocolManager :: URLProtocolManager(
    const char * aProtocolName)
{
  if (aProtocolName == NULL || *aProtocolName == 0)
    throw invalid_argument("bad argument to protocol manager");
  size_t len = strlen(aProtocolName);
  size_t validChars = strspn(aProtocolName, URLProtocolManager_VALID_PROTOCOL_CHARS);
  if (validChars != len)
    throw invalid_argument("protocol name contains invalid characters");

  mProtoName = new char[len+1];
  strcpy(mProtoName, aProtocolName);
  mNext = 0;
}

URLProtocolManager :: ~URLProtocolManager()
{
  delete [] mProtoName;
}

URLProtocolManager*
URLProtocolManager :: findProtocol(const char* protocol,
    const char* url,
    int,
    URLProtocolManager *previous)
{
  char proto[256];
  if (!protocol || !*protocol) {
    // parse the protocol from the URL
    protocol = URLProtocolManager::parseProtocol(proto, sizeof(proto), url);
    if (!protocol || !*protocol)
      return 0;
  }
  URLProtocolManager *p = previous ? previous : mFirstProtocol;
  while(p && strcmp(protocol, p->mProtoName) != 0)
    p = p->mNext;
  return p;
}

URLProtocolHandler*
URLProtocolManager :: findHandler(
    const char* url,
    int flags,
    URLProtocolManager *previous)
{
  URLProtocolManager* mgr = findProtocol(0, url, flags, previous);
  return mgr ? mgr->getHandler(url, flags) : 0;
}

/*
 * Here lie static functions that forward to the right class function in the protocol manager
 */
int
URLProtocolManager :: url_open(void* h, const char* url, int flags)
{
  int retval = -1;

  try
  {
    URLProtocolHandler* handler = URLProtocolManager::getHandler(h);
    retval = handler->url_open(url, flags);
  }
  catch (...)
  {
    retval = -1;
  }
  return retval;
}

URLProtocolHandler *
URLProtocolManager :: getHandler(void *h)
{
  if (!h)
    throw invalid_argument("missing handle");
  return (URLProtocolHandler*)h;
}

int
URLProtocolManager :: url_close(void *h)
{
  int retval = -1;
  try
  {
    URLProtocolHandler* handler = URLProtocolManager::getHandler(h);
    retval = handler->url_close();
    // when we close that handler is no longer valid; you must re-open to re-use.
    delete handler;
  }
  catch (...)
  {
    retval = -1;
  }
  return retval;
}

int
URLProtocolManager :: url_read(void *h, unsigned char* buf, int size)
{
  int retval = -1;
  try
  {
    URLProtocolHandler* handler = URLProtocolManager::getHandler(h);
    retval = handler->url_read(buf, size);
  }
  catch(...)
  {
    retval = -1;
  }
  return retval;
}

int
URLProtocolManager :: url_write(void *h, const unsigned char* buf,
    int size)
{
  int retval = -1;
  try
  {
    URLProtocolHandler* handler = URLProtocolManager::getHandler(h);
    retval = handler->url_write(buf, size);
  }
  catch(...)
  {
    retval = -1;
  }
  return retval;
}

int64_t
URLProtocolManager :: url_seek(void *h, int64_t position,
    int whence)
{
  int64_t retval = -1;
  try
  {
    URLProtocolHandler* handler = URLProtocolManager::getHandler(h);
    retval = handler->url_seek(position, whence);
  }
  catch(...)
  {
    retval = -1;
  }
  return retval;
}

URLProtocolHandler::SeekableFlags
URLProtocolManager :: url_seekflags(void *h, const char* url, int flags)
{
  URLProtocolHandler::SeekableFlags retval = URLProtocolHandler::SK_NOT_SEEKABLE;
  try
  {
    URLProtocolHandler* handler = URLProtocolManager::getHandler(h);
    retval = handler->url_seekflags(url, flags);
  }
  catch(...)
  {
  }
  return retval;
}
}}}}
